File: src/main/java/guru/xgm/language/jass/openapi/actionSystem/convert/jass2angelscript/Jass2AngelScriptVisitor.java
package guru.xgm.language.jass.openapi.actionSystem.convert.jass2angelscript;

import guru.xgm.language.jass.openapi.actionSystem.convert.Jass2AnyVisitor;
import guru.xgm.language.jass.psi.*;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.HashSet;
import java.util.List;
import java.util.Objects;

public class Jass2AngelScriptVisitor extends Jass2AnyVisitor {

    Jass2AngelScriptVisitor() {
        // https://www.angelcode.com/angelscript/sdk/docs/manual/doc_reserved_keywords.html
        keywords = new HashSet<>() {{
            add("abstract");
            add("auto");
            add("bool");
            add("break");
            add("case");
            add("catch");
            add("class");
            add("const");
            add("continue");
            add("default");
            add("do");
            add("double");
            add("enum");
            add("explicit");
            add("external");
            add("final");
            add("float");
            add("for");
            add("from");
            add("funcdef");
            add("get");
            add("import");
            add("in");
            add("inout");
            add("int");
            add("interface");
            add("int8");
            add("int16");
            add("int32");
            add("int64");
            add("is");
            add("mixin");
            add("namespace");
            add("out");
            add("override");
            add("private");
            add("property");
            add("protected");
            add("shared");
            add("super");
            add("switch");
            add("this");
            add("try");
            add("typedef");
            add("uint");
            add("uint8");
            add("uint16");
            add("uint32");
            add("uint64");
            add("void");
            add("while");
            add("xor");
        }};
    }

    @Override
    public @NotNull String getConvertedTypeName(@NotNull String type) {
        return switch (type) {
            case "integer" -> "int";
            case "real" -> "float";
            case "boolean" -> "bool";
            default -> type;
        };
    }

    @Override
    public void appendReal(@NotNull String real) {
        stringBuffer.append(real).append("f");
    }

    // --- typedef https://www.angelcode.com/angelscript/sdk/docs/manual/doc_global_typedef.html

    @Override
    public void appendTypeDef(String name, String base) {
        stringBuffer.append("//typedef ").append(base).append(" ").append(name);
        appendStatementLineEnd();
    }

    // --- variable
    private void appendVarExpr(@Nullable JassExpr expr) {
        if (expr == null) return;
        if (Objects.equals(expr.getText(), "null")) return;
        stringBuffer.append(" = ");
        expr.accept(this);
    }

    @Override
    public void appendVar(boolean constant, boolean global, boolean array, @NotNull String type, String name, @Nullable JassExpr expr) {
        if (constant) stringBuffer.append("const ");
        if (array) {
            stringBuffer.append("array<").append(type).append("> ").append(name);
        } else {
            stringBuffer.append(type).append(" ").append(name);
            appendVarExpr(expr);
        }
        appendStatementLineEnd();
    }

    // --- function
    @Override
    public void appendFunction(@Nullable String returns, String name, @NotNull List<JassParam> params, @NotNull List<JassStmt> statements) {
        if (returns == null) {
            stringBuffer.append("void");
        } else {
            if (returns.equals("code")) {
                stringBuffer.append("BoolexprFunc@");
            } else {
                stringBuffer.append(returns);
            }
        }
        stringBuffer.append(' ').append(name).append("(");
        for (int i = 0; i < params.size(); i++) {
            final var param = params.get(i);
            final var type = param.getTypeName();
            if (type.getCode() != null) {
                stringBuffer.append("BoolexprFunc@");
            } else {
                stringBuffer.append(getConvertedTypeName(type.getText()));
            }
            stringBuffer.append(" ");
            appendSafeName(param.getId().getText());
            if (i < params.size() - 1) stringBuffer.append(", ");
        }
        stringBuffer.append("){\n");
        for (JassStmt stmt : statements) stmt.accept(this);
        stringBuffer.append("}\n");
    }

    // --- statement

    // func as code
    @Override
    public void appendFunctionAsCode(@NotNull String name) {
        stringBuffer.append("@").append(name);
    }

    // if
    @Override
    public void visitIfStmt(@NotNull JassIfStmt o) {
        stringBuffer.append("if (");
        acceptExpr(o.getExpr());
        stringBuffer.append("){\n");
        for (JassStmt stmt : o.getStmtList()) stmt.accept(this);
        stringBuffer.append("}\n");
        for (JassElseIfStmt stmt : o.getElseIfStmtList()) stmt.accept(this);
        for (JassElseStmt stmt : o.getElseStmtList()) stmt.accept(this);
    }

    @Override
    public void visitElseIfStmt(@NotNull JassElseIfStmt o) {
        stringBuffer.append("else if (");
        acceptExpr(o.getExpr());
        stringBuffer.append("){\n");
        for (JassStmt stmt : o.getStmtList()) stmt.accept(this);
        stringBuffer.append("}\n");
    }

    @Override
    public void visitElseStmt(@NotNull JassElseStmt o) {
        stringBuffer.append("else {\n");
        for (JassStmt stmt : o.getStmtList()) stmt.accept(this);
        stringBuffer.append("}\n");
    }

    // set
    @Override
    public void visitSetStmt(@NotNull JassSetStmt o) {
        final var id = o.getId();
        if (id != null) appendSafeName(id.getText());
        final var arr = o.getArrayAccess();
        if (arr != null) arr.accept(this);
        final var expr = o.getExpr();
        if (expr != null) {
            stringBuffer.append(" = ");
            final String s = expr.getText();
            if (s.equals("null")) {
                stringBuffer.append("nil");
            } else {
                expr.accept(this);
            }
        }
        appendStatementLineEnd();
    }

    // loop
    @Override
    public void visitLoopStmt(@NotNull JassLoopStmt o) {
        stringBuffer.append("while (true) {\n");
        for (JassStmt stmt : o.getStmtList()) stmt.accept(this);
        stringBuffer.append("}\n");
    }

    @Override
    public void visitExitWhenStmt(@NotNull JassExitWhenStmt o) {
        final var expr = o.getExpr();
        if (expr == null) return;
        stringBuffer.append("if (");
        expr.accept(this);
        stringBuffer.append(") break;\n");
    }

}

File: src/main/kotlin/raft/war/language/jass/openapi/actionSystem/convert/jass2angelscript/Jass2AngelScriptVisitor.kt
package raft.war.language.jass.openapi.actionSystem.convert.jass2angelscript

import raft.war.language.jass.codeInspection.number.JassRawcode
import raft.war.language.jass.openapi.actionSystem.convert.Jass2AnyVisitor
import raft.war.language.jass.psi.*

/*
ExecuteFunc
array to Dictionary<int,type>
*/
class Jass2AngelScriptVisitor internal constructor() : Jass2AnyVisitor() {
    init {
        // https://www.angelcode.com/angelscript/sdk/docs/manual/doc_reserved_keywords.html
        keywords = object : HashSet<String>() {
            init {
                add("abstract")
                add("auto")
                add("bool")
                add("break")
                add("case")
                add("catch")
                add("class")
                add("const")
                add("continue")
                add("default")
                add("do")
                add("double")
                add("enum")
                add("explicit")
                add("external")
                add("final")
                add("float")
                add("for")
                add("from")
                add("funcdef")
                add("get")
                add("import")
                add("in")
                add("inout")
                add("int")
                add("interface")
                add("int8")
                add("int16")
                add("int32")
                add("int64")
                add("is")
                add("mixin")
                add("namespace")
                add("out")
                add("override")
                add("private")
                add("property")
                add("protected")
                add("shared")
                add("super")
                add("switch")
                add("this")
                add("try")
                add("typedef")
                add("uint")
                add("uint8")
                add("uint16")
                add("uint32")
                add("uint64")
                add("void")
                add("while")
                add("xor")
            }
        }
    }

    override fun getConvertedTypeName(type: String): String {
        return when (type) {
            "integer" -> "int"
            "real" -> "float"
            "boolean" -> "bool"
            else -> type
        }
    }

    override fun appendReal(real: String) {
        stringBuffer.append(real).append("f")
    }

    // --- typedef https://www.angelcode.com/angelscript/sdk/docs/manual/doc_global_typedef.html
    override fun appendTypeDef(name: String, base: String) {
        stringBuffer.append("//typedef ").append(base).append(" ").append(name)
        appendStatementLineEnd()
    }

    // --- variable
    private fun appendVarExpr(expr: raft.war.language.jass.psi.JassExpr?) {
        if (expr == null) return
        if (expr.text == "null") return
        stringBuffer.append(" = ")
        expr.accept(this)
    }

    override fun appendVar(
        constant: Boolean,
        global: Boolean,
        array: Boolean,
        type: String,
        name: String?,
        expr: raft.war.language.jass.psi.JassExpr?
    ) {
        if (constant) stringBuffer.append("const ")
        if (array) {
            stringBuffer.append("array<").append(type).append("> ").append(name)
        } else {
            stringBuffer.append(type).append(" ").append(name)
            appendVarExpr(expr)
        }
        appendStatementLineEnd()
    }

    // --- function
    override fun appendFunction(
        returns: String?,
        name: String?,
        params: List<raft.war.language.jass.psi.JassParam?>,
        statements: List<raft.war.language.jass.psi.JassStmt?>
    ) {
        if (returns == null) {
            stringBuffer.append("void")
        } else {
            if (returns == "code") {
                stringBuffer.append("BoolexprFunc@")
            } else {
                stringBuffer.append(returns)
            }
        }
        stringBuffer.append(' ').append(name).append("(")
        for (i in params.indices) {
            val param = params[i]
            val type = param!!.typeName
            if (type.code != null) {
                stringBuffer.append("BoolexprFunc@")
            } else {
                stringBuffer.append(getConvertedTypeName(type.text))
            }
            stringBuffer.append(" ")
            appendSafeName(param.id.text)
            if (i < params.size - 1) stringBuffer.append(", ")
        }
        stringBuffer.append("){\n")
        for (stmt in statements) stmt!!.accept(this)
        stringBuffer.append("}\n")
    }

    // --- statement
    // func as code
    override fun appendFunctionAsCode(name: String) {
        stringBuffer.append("@").append(name)
    }

    // if
    override fun visitIfStmt(o: raft.war.language.jass.psi.JassIfStmt) {
        stringBuffer.append("if (")
        acceptExpr(o.expr)
        stringBuffer.append("){\n")
        for (stmt in o.stmtList) stmt.accept(this)
        stringBuffer.append("}\n")
        for (stmt in o.elseIfStmtList) stmt.accept(this)
        for (stmt in o.elseStmtList) stmt.accept(this)
    }

    override fun visitElseIfStmt(o: raft.war.language.jass.psi.JassElseIfStmt) {
        stringBuffer.append("else if (")
        acceptExpr(o.expr)
        stringBuffer.append("){\n")
        for (stmt in o.stmtList) stmt.accept(this)
        stringBuffer.append("}\n")
    }

    override fun visitElseStmt(o: raft.war.language.jass.psi.JassElseStmt) {
        stringBuffer.append("else {\n")
        for (stmt in o.stmtList) stmt.accept(this)
        stringBuffer.append("}\n")
    }

    // set
    override fun visitSetStmt(o: raft.war.language.jass.psi.JassSetStmt) {
        val id = o.id
        if (id != null) appendSafeName(id.text)
        val arr = o.arrayAccess
        arr?.accept(this)
        val expr = o.expr
        if (expr != null) {
            stringBuffer.append(" = ")
            val s = expr.text
            if (s == "null") {
                stringBuffer.append("nil")
            } else {
                expr.accept(this)
            }
        }
        appendStatementLineEnd()
    }

    // loop
    override fun visitLoopStmt(o: raft.war.language.jass.psi.JassLoopStmt) {
        stringBuffer.append("while (true) {\n")
        for (stmt in o.stmtList) stmt.accept(this)
        stringBuffer.append("}\n")
    }

    override fun visitExitWhenStmt(o: raft.war.language.jass.psi.JassExitWhenStmt) {
        val expr = o.expr ?: return
        stringBuffer.append("if (")
        expr.accept(this)
        stringBuffer.append(") break;\n")
    }


    // --- expressions
    override fun appendRawcode(raw: JassRawcode) {
        if (raw.safe) stringBuffer.append("'").append(raw.strval).append("'")
        else stringBuffer.append(raw.hex)
    }
}
